File: generator/src/test/java/com/cromoteca/generator/nullability/BasicEndpoint.java
package com.cromoteca.generator.nullability;

import dev.hilla.Endpoint;
import dev.hilla.Nonnull;
import java.util.List;
import java.util.Map;

@Endpoint
public class BasicEndpoint {
  public Integer box(int n) {
    return n;
  }

  @Nonnull
  public String notNull() {
    return "";
  }

  public void nulls(Nulls argument) {}

  public void nonNullListOfNonNullStrings(@Nonnull List<@Nonnull String> argument) {}

  public static class Nulls {
    private @Nonnull Map<@Nonnull String, @Nonnull List<@Nonnull String>> allNull;
    private @Nonnull Map<@Nonnull String, @Nonnull List<String>> someNull;
    private @Nonnull List<@Nonnull Object> nullOrNot;

    public Map<String, List<String>> getAllNull() {
      return allNull;
    }

    public void setAllNull(@Nonnull Map<String, List<String>> allNull) {
      this.allNull = allNull;
    }

    public Map<String, List<String>> getSomeNull() {
      return someNull;
    }

    public void setSomeNull(Map<String, List<String>> someNull) {
      this.someNull = someNull;
    }

    // The type argument is nullable and cannot inherit the annotation since the type argument of
    // the field is different.
    public @Nonnull List<String> getNullOrNot() {
      return nullOrNot.stream().map(Object::toString).toList();
    }
  }
}

File: generator/src/test/resources/com/cromoteca/generator/nullability/BasicEndpoint.ts
import client from '../../../../connect-client.default.js';
import type { EndpointRequestInit } from '@hilla/frontend';

async function box(n: number, init?: EndpointRequestInit): Promise<number | undefined> {
    return client.call('com.cromoteca.generator.nullability.BasicEndpoint', 'box', { n }, init);
}

async function nonNullListOfNonNullStrings(argument: string[], init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.generator.nullability.BasicEndpoint', 'nonNullListOfNonNullStrings', { argument }, init);
}

async function notNull(init?: EndpointRequestInit): Promise<string> {
    return client.call('com.cromoteca.generator.nullability.BasicEndpoint', 'notNull', {}, init);
}

const BasicEndpoint = {
    box,
    nonNullListOfNonNullStrings,
    notNull,
};

export default BasicEndpoint;
